src/dataflow/src/render/reduce.rs

```diff
@@ -7,6 +7,51 @@
+//! 1. Accumulable:
+//!    Accumulable reductions can be computed inline in a Differential update's `difference`
+//!    field because they basically boil down to tracking counts of things. `sum()` is an
+//!    example of an accumulable reduction, and when some element `x` is removed from the set
+//!    of elements being summed, we can introduce `-x` to incrementally maintain the sum. More
+//!    formally, accumulable reductions correspond to instances of Abelian groups.
+//! 2. Hierarchical:
+//!    Hierarchical reductions don't have a meaningful negation like accumulable reductions do, but
+//!    they are still associative, which lets us compute the reduction over subsets of the
+//!    input, and then compute the reduction again on those results. For example:
+//!    `min[2, 5, 1, 10]` is the same as `min[ min[2, 5], min[1, 10]]`. When we compute hierarchical
+//!    reductions this way, we can maintain the computation in sublinear time with respect to
+//!    the overall input. `min` and `max` are two examples of hierarchical reductions. More formally,
+//!    hierarchical reductions correspond to instances of semigroups, in that they are associative,
+//!    but in order to benefit from being computed hierarchically, they need to have some reduction
+//!    in data size as well. A function like "concat-everything-to-a-string" wouldn't benefit from
+//!    hierarchical evaluation.
+//!
+//!    When the input is append-only, or monotonic, reductions that would otherwise have to be computed
+//!    hierarchically can instead be computed in-place, because we only need to keep the value that's
+//!    better than the "best" (minimal or maximal for min and max) seen so far.
+//! 3. Basic:
+//!    Basic reductions are a bit like the Hufflepuffs of this trifecta. They are neither accumulable nor
+//!    hierarchical (most likely they are associative but don't involve any data reduction) and so for these
+//!    we can't do much more than just defer to Differential's reduce operator and eat a large maintenance cost.
+//!
+//! When render these reductions we want to limit the number of arrangements we produce. Therefore, if we only
+//! have multiple instances of a single reduction type in our reduce expression, we'll specialize and only
+//! render a dataflow to compute those functions. If instead we have instances of multiple reduction types
+//! in the same expression, we'll need to divide them up by type, render them separately, and then take those
+//! results and collate them back in the requested order.
+//!
+//! We build `ReducePlan`s to manage the complexity of planning the generated dataflow for a given reduce
+//! expression. The intent here is that each creating a `ReducePlan` should capture all of the decision making
+//! about what kind of dataflow do we need to render and what each operator needs to do, and then actually
+//! rendering the plan can be a relatively simple application of (as much as possible) straight line code.
+
```

Comment:
This is the important para, imo. If you want to lift this up to the first paragraph I would not object!